src/legacy/core_plugins/elasticsearch/index.js

```diff
@@ -17,79 +17,86 @@
-        startupTimeout: Joi.number().default(5000),
-        logQueries: Joi.boolean().default(false),
-        ssl: sslSchema,
-        apiVersion: Joi.string().default('master'),
-        healthCheck: Joi.object({
-          delay: Joi.number().default(2500)
-        }).default(),
-      }).default();
-    },
-
-    uiExports: {
-      injectDefaultVars(server, options) {
-        return {
-          esRequestTimeout: options.requestTimeout,
-          esShardTimeout: options.shardTimeout,
-          esApiVersion: options.apiVersion,
-        };
-      }
-    },
-
-    init(server) {
-      const clusters = createClusters(server);
-
-      server.expose('getCluster', clusters.get);
-      server.expose('createCluster', clusters.create);
-
-      server.expose('filterHeaders', filterHeaders);
-      server.expose('ElasticsearchClientLogging', clientLogger(server));
-
-      createDataCluster(server);
-      createAdminCluster(server);
+
+    uiExports: { injectDefaultVars: () => defaultVars },
+
+    async init(server) {
+      // All methods that ES plugin exposes are synchronous so we should get the first
+      // value from all observables here to be able to synchronously return and create
+      // cluster clients afterwards.
+      const [esConfig, adminCluster, dataCluster] = await combineLatest(
+        server.core.elasticsearch.legacy.config$,
+        server.core.elasticsearch.adminClient$,
+        server.core.elasticsearch.dataClient$
+      ).pipe(
+        first(),
+        map(([config, adminClusterClient, dataClusterClient]) => [
+          config,
+          new Cluster(adminClusterClient),
+          new Cluster(dataClusterClient)
+        ])
+      ).toPromise();
+
+      defaultVars = {
+        esRequestTimeout: esConfig.requestTimeout.asMilliseconds(),
+        esShardTimeout: esConfig.shardTimeout.asMilliseconds(),
+        esApiVersion: esConfig.apiVersion,
+      };
+
+      const clusters = new Map();
+      server.expose('getCluster', (name) => {
+        if (name === 'admin') {
+          return adminCluster;
+        }
+
+        if (name === 'data') {
+          return dataCluster;
+        }
+
+        return clusters.get(name);
+      });
+
+      server.expose('createCluster', (name, clientConfig = {}) => {
+        if (clusters.has(name)) {
+          throw new Error(`cluster '${name}' already exists`);
+        }
+
+        // We fill all the missing properties in the `clientConfig` using the default
+        // Elasticsearch config so that we don't depend on default values set and
+        // controlled by underlying Elasticsearch JS client.
+        const cluster = new Cluster(server.core.elasticsearch.createClient(name, {
+          ...esConfig,
+          ...clientConfig,
+        }));
+
+        clusters.set(name, cluster);
+
+        return cluster;
+      });
+
+      server.events.on('stop', () => {
+        for (const cluster of clusters.values()) {
+          cluster.close();
```

Comment:
> seems that we should pre-fill `clusters` map with 'admin' & 'data', otherwise we don't stop them.

Core's admin and data clusters should be stopped by the core/new platform when Elasticsearch service is stopped. All other cluster clients created through the legacy shim should be put into this map and stopped when legacy world is stopped (`stop` event on server).

> and another corner case: `createCluster` allows to create new clusters with 'admin', 'data' types

I left this escape hatch intentionally to support custom `admin` client created by the monitoring plugin (see in this PR and https://github.com/elastic/kibana/blob/master/x-pack/plugins/monitoring/server/kibana_monitoring/bulk_uploader.js#L51-L53). It's bad even though not harmful right now (as it won't be returned by `getCluster`) and we should remove this "feature" as soon as we find solution for monitoring. Let me know if you have any thoughts on that, would be glad to fix this, otherwise my plan is to leave it as is (with comment) for now and tackle separately.